allennlp/data/dataset_readers/sharded_dataset_reader.py

```diff
@@ -0,0 +1,50 @@
+import glob
+import logging
+from typing import Iterable
+
+from allennlp.data.dataset_readers.dataset_reader import DatasetReader
+from allennlp.data.instance import Instance
+
+
+logger = logging.getLogger(__name__)
+
+
+@DatasetReader.register("sharded")
+class ShardedDatasetReader(DatasetReader):
```

Comment:
Yeah, at the moment this is very much for testing/experimentation. I don't know what form it will take in the long term, i.e. by 1.0. I think that will be informed in part by your iterator work, and perhaps vice versa. My current thoughts are as follows:

1) Allowing globbing, separate from selecting shards for use in the distributed case, is a useful feature. I'd like that to exist in the main library. No opinion on where. At the moment only `MultiprocessDatasetReader` allows this and I'm not using that because it doesn't scale linearly, to avoid flakiness, to avoid serialization overhead etc.

2) Sharding is very standard in distributed systems. To the extent that we want to support large datasets and multiple GPUs it seems natural for it to live in the main library.

3) Modulo `num_workers` is fine for small non-sharded datasets that don't do lots of generation in the reader. Happy to support that too. (I think this works in the lazy case as well?) Regardless, I'll need something like `ShardedDatasetReader` for the 1 billion word corpus.